### PR TITLE
Create cohort

### LINKED ISSUE #26: Add upload classes

### PR DESCRIPTION
**Pull request by theodorusandi**
closes #24
closes #26

### CODE REVIEW

File: controllers/reset-password.controller.js [hunk 1]
```diff
@@ -17,28 +17,24 @@ function ResetPasswordController() {
...
-        res.set(400).send({ msg: "fail to update." });
+        res.set(400).send({ msg: 'fail to update.' });
       }
 
-      const deleted = await passwordResetTokens.deleteTokenBy("email", email);
+      const deleted = await passwordResetTokens.deleteTokenBy('email', email);
 
       if (deleted.deletedCount !== 1) {
-        res.set(400).send({ msg: "fail to update." });
+        res.set(400).send({ msg: 'fail to update.' });
```
Comment (RDiMag) [inline controllers/reset-password.controller.js]: I'd suggest this change: 'Failed to update.'

File: middlewares/requests/admin/admin-class-store.request.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+const adminClassStoreRequest = [
+  body('name').isLength({ min: 1, max: 255 }).trim().escape(),
+  body('subject').isLength({ min: 1, max: 255 }).trim().escape(),
+  param('cohortId').custom(async (value) => {
+    try {
+      console.log(value);
+      const cohort = await cohorts.findCohortById(value);
+      if (!cohort) {
+        // eslint-disable-next-line prefer-promise-reject-errors
+        return Promise.reject('cohort ID is invalid');
```
Comment (RDiMag) [inline middlewares/requests/admin/admin-class-store.request.js]: Recommended change: 'Cohord ID is invalid.'

File: middlewares/requests/reset-password.request.js [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
     .isEmail()
     .isLength({ min: 1, max: 255 })
     .normalizeEmail({ gmail_remove_dots: false }),
-  body("password").isAlphanumeric().isLength({ min: 1, max: 255 }).trim(),
-  body("passwordConfirmation").custom((value, { req }) => {
+  body('password').isAlphanumeric().isLength({ min: 1, max: 255 }).trim(),
+  body('passwordConfirmation').custom((value, { req }) => {
     if (value !== req.body.password) {
-      throw new Error("Password confirmation does not match password");
+      throw new Error('Password confirmation does not match password');
```
Comment (RDiMag) [inline middlewares/requests/reset-password.request.js]: Suggestion: 'Password confirmation does not match password.'

File: server.js [hunk 4]
```diff
@@ -50,9 +57,9 @@ app.use(router);
 
 // error handler
 app.use(function (err, req, res, next) {
-  if (err.code === "EBADCSRFTOKEN") {
+  if (err.code === 'EBADCSRFTOKEN') {
     // handle CSRF token errors here
-    return res.status(403).send({ message: "form tampered with" });
+    return res.status(403).send({ message: 'form tampered with' });
```
Comment (RDiMag) [inline server.js]: Recommended: 'Form has been tampered with.'

**Review by RDiMag — APPROVED** [review]

File: controllers/admin/admin-class.controller.js [hunk 5]
```diff
@@ -1,7 +1,30 @@
+const classes = require('../../dao/class.dao');
+
 function AdminClassController() {
```
Comment (jdmedlock) [inline controllers/admin/admin-class.controller.js:3]: This is just a suggestion, not a requirement. The JS community has and is shifting to using arrow notation. Evaluate whether using function definitions like `this.index2 = (req, res, next) => {...}`. This doesn't change how the function works or it's performance - it's just a more modern way to define it.

File: controllers/admin/admin-cohort.controller.js [hunk 6]
```diff
@@ -1,9 +1,27 @@
...
+      name,
+      startDate,
+      endDate,
+    };
+
+    try {
+      const result = await cohorts.insertCohort(cohort);
+      res.status(200).send({ msg: 'success!', data: result });
+    } catch (err) {
+      res.status(500).send({ msg: "couldn't insert cohort" });
```
Comment (jdmedlock) [inline controllers/admin/admin-cohort.controller.js]: Is there any other information that if included in the error message would help in debugging. For example, a cohort name or id>

File: controllers/auth-github.controller.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        return next(err);
+      }
+      if (!user) {
+        res.status(401).send({ error: { message: info.message } });
+      }
+      req.logIn(user, (err) => {
+        if (err) {
+          return next(err);
+        }
+        res.redirect('http://localhost:3000/admin/cohorts');
```
Comment (jdmedlock) [inline controllers/auth-github.controller.js]: Consider adding this URL as a configuration option in `.env` so it's easier to customize for the Production vs. Development environment

File: dao/classwork-type.dao.js [hunk 8]
```diff
@@ -19,8 +19,9 @@ function ClassworkTypeDao() {
       return cursor.toArray();
```
Comment (jdmedlock) [inline dao/classwork-type.dao.js:19]: The approach of converting return values to the simplest form is a very good practice. Doing this in the function that's exposed to the details insulates the calling function from being exposed to the details of a particular data structure. This removes an implicit dependency and makes the code easier to maintain & enhance.

File: dao/password-reset-token.dao.js [hunk 9]
```diff
@@ -7,12 +7,9 @@ function PasswordResetTokenDao() {
     }
 
     try {
-      passwordResetTokens = await client.db().collection("passwordResetTokens");
+      passwordResetTokens = await client.db().collection('passwordResetTokens');
 
-      passwordResetTokens.createIndex(
-        { createdAt: 1 },
```
Comment (jdmedlock) [inline dao/password-reset-token.dao.js:13]: Be cautious about "magic" numbers. Something I like to do is create constants with a name that defines what the number represents. This makes the code more readable and if the magic number is used in multiple locations in the code it allows you to change it by updating a single line of code. For example, `{ createdAt: defaultCreationTS }` or `{ expireAfterSeconds: defaultExpirationTime }`

File: dao/user.dao.js [hunk 10]
```diff
@@ -29,7 +29,7 @@ function UserDao() {
       isAdmin: false,
       createdAt: new Date(),
       photo:
-        "https://images.unsplash.com/photo-1623584973952-182bcb43b8ad?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=480&q=80",
```
Comment (jdmedlock) [inline dao/user.dao.js:32]: Consider defining this as a constant so you can use a name that provides contextual meaning. You might even want to define this in `.env` or a config file to make it easy to change one image for another. For example, `photo: errorUserImage,`.

File: middlewares/passport.middleware.js [hunk 11]
```diff
@@ -1,46 +1,70 @@
...
+        return cb(null, false, {
+          message: 'Incorrect username or password.',
+        });
+      }
+
+      const valid = validatePassword(password, user.password, user.salt);
+
+      if (!valid) {
+        return cb(null, false, {
+          message: 'Incorrect username or password.',
```
Comment (jdmedlock) [inline middlewares/passport.middleware.js]: Something I like to do on the server is to also issue a console.log to record the error and any related information (like a user id). This makes it easier to get the info you need for debugging if a user reports an issue hours or days after it occurred.

File: middlewares/passport.middleware.js [hunk 12]
```diff
@@ -1,46 +1,70 @@
...
+    }
+  })
+);
+
+passport.use(
+  new GitHubStrategy(
+    {
+      clientID: process.env.GITHUB_CLIENT_ID,
+      clientSecret: process.env.GITHUB_CLIENT_SECRET,
+      callbackURL: 'http://localhost:5000/auth/github/callback',
```
Comment (jdmedlock) [inline middlewares/passport.middleware.js]: Consider adding this as a `.env` variable. See previous comments for reasons why this is a good practice.

File: middlewares/requests/admin/admin-class-store.request.js (hunk 2 above)
Comment (jdmedlock) [reply middlewares/requests/admin/admin-class-store.request.js]: Sorry for being picky. This should be 'Cohort ID is invalid.'

**Review by jdmedlock — APPROVED** [review]
See individual comments I've added. These are all suggestions and things to think about rather than required changes. I'm impressed with the fact you've spent a lot of time handling edge cases and intercepting errors!

File: middlewares/requests/admin/admin-class-store.request.js (hunk 2 above)
Comment (theodorusandi) [reply middlewares/requests/admin/admin-class-store.request.js]: LOL

**Pull request merged**